tests/test_cases/test_axi4/test_axi4.py

```diff
@@ -0,0 +1,451 @@
+                                 AXIxRESP)
+from cocotb.regression import TestFactory
+from cocotb.result import TestFailure
+from cocotb.triggers import ClockCycles, Combine, Join, RisingEdge
+
+
+CLK_PERIOD = (10, "ns")
+AXI_PREFIX = "S_AXI"
+
+
+def get_parameters(dut):
+    address_width = dut.ADDR_WIDTH.value // 8
+    data_width = dut.DATA_WIDTH.value // 8
+    ram_start = dut.RAM_BASE_ADDRESS.value
+    ram_stop = ram_start + 2**dut.RAM_WIDTH.value
+    return address_width, data_width, ram_start, ram_stop
+
+
+def add_wstrb_mask(data_width, previous_value, write_value, wstrb):
+    result = 0
+    for i in range(data_width):
+        source = write_value if wstrb & (1 << i) else previous_value
+        result |= (source & (0xff << (i * 8)))
+
+    return result
+
+
+async def setup_dut(dut):
+    cocotb.fork(Clock(dut.clk, *CLK_PERIOD).start())
+    dut.rstn <= 0
+    await ClockCycles(dut.clk, 2)
+    dut.rstn <= 1
+    await ClockCycles(dut.clk, 2)
+
+
+async def test_single_beat(dut, driver, address_latency, data_latency):
+    """Test single read/write"""
+
+    axim = driver(dut, AXI_PREFIX, dut.clk)
+    _, data_width, ram_start, ram_stop = get_parameters(dut)
+
+    await setup_dut(dut)
+
+    address = randrange(ram_start, ram_stop, data_width)
+    write_value = randrange(0, 2**(data_width * 8))
+    strobe = randrange(1, 2**data_width)
+
+    previous_value = await axim.read(address)
+    await axim.write(address, write_value, byte_enable=strobe,
+                     address_latency=address_latency,
+                     data_latency=data_latency)
+    read_value = await axim.read(address)
+
+    if isinstance(read_value, list):
+        previous_value = previous_value[0]
+        read_value = read_value[0]
+
+    expected_value = add_wstrb_mask(data_width, previous_value.integer,
+                                    write_value, strobe)
+
+    if read_value != expected_value:
+        raise TestFailure("Read {:#x} from {:#x} but was expecting {:#x} "
+                          "({:#x} with {:#x} as strobe and {:#x} as previous "
+                          "value)"
+                          .format(read_value.integer, address, expected_value,
+                                  write_value, strobe, previous_value))
+
+
+async def test_incr_burst(dut, return_rresp):
+    """Test burst reads/writes"""
+
+    axim = AXI4Master(dut, AXI_PREFIX, dut.clk)
+    _, data_width, ram_start, ram_stop = get_parameters(dut)
+
+    await setup_dut(dut)
+
+    burst_length = randrange(2, 256)
+    base = randrange(ram_start, ram_stop, 4096)
+    offset = randrange(0, 4096 - (burst_length - 1) * data_width, data_width)
+    address = base + offset
+    write_values = \
+        [randrange(0, 2**(data_width * 8)) for i in range(burst_length)]
+
+    # Make strobe one item less than data, to test also that driver behavior
+    # This RAM implementation does not honor WSTRB=0
```

Comment:
This should not be the case, WSTRB=0 should be handled correctly by axi_ram.v

Reply:
I had this issue where all the WSTRB were handled correctly, except for all WSTRB=0 which was treated as all ones. However I can not reproduce it anymore, so I do not know if it was a mistake in my test code or not.

As it does not appear here, I will use the full WSTRB range in this test.